tsec-http4s/src/main/scala/tsec/authentication/JWKPublicKeyRSAAuthenticator.scala

```diff
@@ -0,0 +1,197 @@
+    getKey(id).map {
+      case None if shouldFetch() =>
+        for {
+          jwks      <- fetchJwks(uri)
+          k         <- E.delay(jwks.keys.map(jwk => (jwk.kid, convert(jwk))).toMap)
+          _         <- E.delay(keys.set(k))
+          _         <- E.delay(lastFetch.set(Instant.now().some))
+          key       <- E.delay(k.get(id))
+        } yield key
+      case _ => E.delay(keys.get().get(id))
+    }.flatten
+  }
+
+  private def fetchJwks(uri:Uri) = {
+    val s = for {
+      client <- Http1Client.stream[F]()
+      jwks <- Stream.eval(client.expect[JWKS](uri))
+    } yield jwks
+    s.compile.last.flatMap {
+      case Some(jwks) => jwks.pure
+      case None => E.raiseError[JWKS](new Exception("Error fetching JWKS"))
+    }
+  }
+
+  private def shouldFetch() =
+    lastFetch.get().forall(_.isBefore(Instant.now().minusSeconds(minFetchDelay.toSeconds)))
+
+  private def getKey(id: String) = E.delay {
+    keys.get().get(id)
+  }
+
+  private def convert(jwk: JWK): SigPublicKey[A] = {
+    val pubKey = KeyFactory.getInstance(jwk.kty).generatePublic(new RSAPublicKeySpec(jwk.n.value.bigInteger, jwk.e.value.bigInteger))
+    SigPublicKey[A](pubKey)
+  }
+
+}
+
+class JWKPublicKeyRSAAuthenticator[F[_] : Effect, I: Decoder, V, A: JCASigTag](
+   expiryDuration: FiniteDuration,
+   maxIdleDuration: Option[FiniteDuration],
+   identityStore: IdentityStore[F, I, V],
+   jwksUri: Uri,
+   minFetchDelay: FiniteDuration
+)(implicit cv: JWSSigCV[F, A]) extends Authenticator[F, I, V, AugmentedJWK[A, I]] {
+
+  private val keyRegistry = new KeyRegistry[F, A](jwksUri, minFetchDelay)
+
+  override def expiry: FiniteDuration = expiryDuration
+
+  override def maxIdle: Option[FiniteDuration] = None
+
+  /** Attempt to retrieve the raw representation of an A
+    * This is primarily useful when attempting to combine AuthenticatorService,
+    * to be able to evaluate an endpoint with more than one token type.
+    * or simply just to prod whether the request is malformed.
+    *
+    * @return
+    */
+  override def extractRawOption(request: Request[F]): Option[String] =
+    request.headers.get(Authorization).flatMap { t =>
+      t.credentials match {
+        case Credentials.Token(scheme, token) if scheme == AuthScheme.Bearer =>
+          Some(token)
+        case _ => None
+      }
+    }
+
+  /** Parse the raw representation from `extractRawOption`
+    *
+    */
+  override def parseRaw(raw: String, request: Request[F]): OptionT[F, SecuredRequest[F, V, AugmentedJWK[A, I]]] =
+    OptionT(
+      (for {
+        now          <- Sync[F].delay(Instant.now())
+        extractedRaw <- cv.extractRaw(raw)
+        publicKey    <- keyRegistry.getPublicKey(extractedRaw.header.kid.get)
+        extracted    <- cv.verify(raw, publicKey.get, now)
+        jwtid        <- cataOption(extracted.body.subject)
+        id           <- cataOption(extracted.body.subject.flatMap(s => decode[I](s""""$s"""").toOption))
```

Comment:
Why are you doing this? you can just write `decode[I](_).toOption`

Reply:
When `I` is `String` the value passed to `decode` needs to be quoted. Otherwise Circe expects a json value:

```
scala> import io.circe.parser.decode
import io.circe.parser.decode

scala> decode[String]("user1")
res1: Either[io.circe.Error,String] = Left(io.circe.ParsingFailure: expected json value got u (line 1, column 1))

scala> decode[String](s""""user1"""")
res2: Either[io.circe.Error,String] = Right(user1)
```

Reply:
yes, this is for a _string_ value, not a json object.

Reply:
Wouldn't this also cause problems in e.g. [PartialStatelessJWTAuth](https://github.com/jmcardon/tsec/blob/194c26f3b45a1d5292faee5bea7fd4f496713fcf/tsec-http4s/src/main/scala/tsec/authentication/internal/PartialStatelessJWTAuth.scala#L40)?

If I have an `IdentityStore[F[_], I, V]` and `I` is `String` the decoding `extracted.body.subject.flatMap(decode[I](_).toOption)` would always fail, no?

Reply:
I think I'd be more concerned with someone having their identity type be a string in the first place.

Reply:
If I have a value class representing my user identity and using `deriveUnwrappedDecoder` it would also fail, wouldn't it?

Reply:
if it's correctly supplied by the user, the string itself will come with quotes. We should not be adding this ourselves.

Reply:
Are you saying the `sub` field in the JWT should have quotes (e.g. `"sub": "\"user1\"",`)? Sorry for being a pain in the arse. Just want to understand how it is intended to work.

Reply:
Also, if you are using something like auth0 you don't have much influence on this.

Reply:
I changed this as you suggested. But still interested how to make tsec work for my use case where the `sub` is not in my control.